### ISSUE TITLE
Is it possible to use refit with HttpClientFactory in .net framework winform application?

### ISSUE THREAD

**Issue by softboy99** [reactiveui/refit#1004]
Hi,
I'm not familiar with HttpClientFactory,which appears in asp.net core 2.1 above,Is it possible to use refit with HttpClientFactory in .net framework winform application?

**Comment by nicostouch** [comment]
HttpClientFactory is only part of .NET from .NET Core 2.1 and onwards, so .NET 4.x doesn't have it. Refit has full support for HttpClientFactory / DependencyInjection, but doesn't require it. Refit works by making http requests with HttpClient which is part of .NET 4.5, so any version of .NET >= 4.5 is compatible with Refit. 

It just means you'll need to use the `RestService.For<T>` semantics. If your application uses some Depedency Injection container like Autofac, Ninject or LightInject etc you can simply register a Refit client using your DI container with the RestService.For<T>. In .NET Core the Refit's support for HttpClientFactory is just doing that same thing for you basically. 

In .NET Core 2.1+ all HttpClientFactory really does is manage the lifecycle of HttpClient objects a bit more intelligently to avoid some problems with HttpClient in older versions of .NET Framework. 

Does that answer your question?

**Comment by softboy99** [comment]
Hi
It seems that, in .net framework, we also can use HttpClientFactory. Just adding Microsoft.Extensions.Http and Microsoft.Extensions.DependencyInjection, Could you please kindly help to give some sample code for that? Thanks a lot for your great help.

**Comment by james-s-tayler** [comment]
True. I forgot it's technically netstandard2.0 compatible. The .NET 4.x project I work on isn't really in a position to upgrade to even make use of `netstandard2.0` stuff, so I tend to equate `netstandard2.0` exclusively with .NET Core in my mind. 

Getting HttpClientFactory to work in .NET 4.x isn't a Refit specific question, so you might have better luck asking on Stack Overflow for that. 

On the flipside - does your project actually require the use of HttpClientFactory? It only really helps solve the problems of DNS updates not being respected by HttpClient and socket exhaustion from instantiating too many HttpClients that keep the sockets open for a while after the request completes. I'd imagine in a winforms application this might not be a problem?

Anyway, good luck.

**Comment by softboy99** [comment]
Hi,
I know how to do that in a normal .net framework winform application with httpclientfactory, but i don't know how to integrate it with your refit be configured together with httpclientfactory. Your documentation only for asp.net core. Could you please kindly help me? Thanks a lot.

**Comment by james-s-tayler** [comment]
If you have already got `HttpClientFactory` working in your project using the `IServiceCollection` interface from the `netstandard2.0` dependency injection then it should be the same as the docs using the AddRefitClient<T> extension method on `IServiceCollection`. 

If you've gotten `HttpClientFactory` working in your project without somehow making use of `IServiceCollection` then you'll have to provide some details on what you've done, or what you have tried so far.

**Comment by softboy99** [comment]
Hi,
i don't know how to use IServiceCollection  interface in winform with .net framework, IServiceCollection  interface is in asp.net

**Comment by ChrisPulman** [comment]
You could see if Dapplo.Microsoft.Extensions.Hosting.ReactiveUI.WinForms works for you.

**Comment by glennawatson** [comment]
Refit works fine on .NET Framework / WinForms without HttpClientFactory - just use `RestService.For<IMyApi>("https://...")` or `RestService.For<IMyApi>(httpClient, settings)` and manage the `HttpClient` lifetime yourself (a single long-lived instance is fine for most desktop apps).

If you do want `AddRefitClient<T>` + `HttpClientFactory`, it requires `Microsoft.Extensions.DependencyInjection` / `Microsoft.Extensions.Http` and an `IServiceCollection`; that is not Refit-specific setup and works the same way once you have a service collection (e.g. via Generic Host). That integration question is better suited to Discussions or Stack Overflow.

Routing this to https://github.com/reactiveui/refit/discussions and closing.


<blockquote><img src="https://opengraph.githubassets.com/7da521d756b94831ff31d8d5fc75ecbe2c2d69a69928da79595847192e775278/reactiveui/refit" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/reactiveui/refit/discussions">reactiveui refit · Discussions</a></strong></div><div>Explore the GitHub Discussions forum for reactiveui refit. Discuss code, ask questions & collaborate with the developer community.</div></blockquote>
